src/Symfony/Component/DependencyInjection/Loader/FileLoader.php

```diff
@@ -156,12 +157,25 @@ public function registerClasses(Definition $prototype, string $namespace, string
                 }
                 if ($this->env) {
                     $attribute = null;
-                    foreach ($r->getAttributes(When::class, \ReflectionAttribute::IS_INSTANCEOF) as $attribute) {
-                        if ($this->env === $attribute->newInstance()->env) {
+                    $whenAttributes = $r->getAttributes(When::class, \ReflectionAttribute::IS_INSTANCEOF);
+                    $notWhenAttributes = $r->getAttributes(NotWhen::class, \ReflectionAttribute::IS_INSTANCEOF);
+
+                    if ($whenAttributes && $notWhenAttributes) {
+                        throw new LogicException(sprintf('The "%s" class cannot have both #[When] and #[NotWhen] attributes.', $class));
+                    }
+
+                    $attributes = $whenAttributes ?: $notWhenAttributes;
+                    foreach ($attributes as $attribute) {
+                        $instance = $attribute->newInstance();
+                        if (
+                            $instance instanceof When && $this->env === $instance->env
+                            || $instance instanceof NotWhen && $this->env !== $instance->env
```

Comment:
this implementation is wrong for `NotWhen`. You cannot accept defining a service as soon as you find one `NotWhen` attribute with a different attribute (as another `NotWhen` could be forbidding it). the various `NotWhen` attributes must be combined in a `AND` fashion, not a `OR` fashion